### ISSUE TITLE
Investigate potential perf improvements in `colour-science` codebase

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#9109]
See [this comment](https://github.com/python/mypy/issues/12225#issuecomment-2381608363).

On initial inspection, it appears that the file `colour/characterisation/datasets/colour_checkers/chromaticity_coordinates.py` is taking excessive time (55sec on a fast machine) to type check. Other files are also taking much longer than I'd expect.

```
55124ms: /characterisation/datasets/colour_checkers/chromaticity_coordinates.py
7455ms: /notation/datasets/munsell/all.py
3773ms: /notation/datasets/munsell/real.py
3693ms: /geometry/tests/test_vertices.py
3103ms: /models/rgb/tests/test_rgb_colourspace.py
2597ms: /utilities/tests/test_array.py
2117ms: /io/luts/tests/test_lut.py
1958ms: /models/tests/test_cie_luv.py
1625ms: /corresponding/tests/test_prediction.py
1524ms: /continuous/tests/test_multi_signal.py
1262ms: /notation/tests/test_munsell.py
1255ms: /notation/datasets/munsell/experimental.py
1252ms: /quality/tests/test_cqs.py
1240ms: /models/tests/test_cie_xyy.py
1158ms: /geometry/tests/test_primitives.py
1156ms: /characterisation/tests/test_aces_it.py
1089ms: /algebra/tests/test_common.py
1077ms: /models/rgb/tests/test_ictcp.py
1037ms: /models/tests/test_cie_ucs.py
1026ms: /colorimetry/tests/test_tristimulus_values.py
1009ms: /models/rgb/tests/test_derivation.py
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.383.

**Comment by erictraut** [comment]
I needed to back out the performance optimization because it caused a regression in protocol matching. I'll need to investigate other ways to address this.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.384
